### PR TITLE
mon/FSCommands: add command to rename a file system

### PR DESCRIPTION
**Pull request by ajarr**
Fixes: https://tracker.ceph.com/issues/47276
Signed-off-by: Ramana Raja <rraja@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ajarr** [comment]
requires more testing and need to add  a test case

File: src/mon/FSCommands.cc [hunk 1]
```diff
@@ -968,6 +968,56 @@ class ResetFilesystemHandler : public FileSystemCommandHandler
...
+        ss << "filesystem '" << fs_name << "' does not exist";
+        // Unlike fs rm, we consider this case an error
+        return -ENOENT;
+    }
+
+    // check new filesystem doesn't already exist!
+    string new_fs_name;
+    cmd_getval(cmdmap, "new_fs_name", new_fs_name);
+    auto new_fs = fsmap.get_filesystem(new_fs_name);
+    if (!(new_fs == nullptr)) {
```
Comment (batrick) [inline src/mon/FSCommands.cc:998]:
```suggestion
    if (new_fs) {
```

File: src/mon/FSCommands.cc [hunk 2]
```diff
@@ -968,6 +968,56 @@ class ResetFilesystemHandler : public FileSystemCommandHandler
...
+        ss << "desired filesystem name '" << new_fs_name << "' already in use";
+        return -EINVAL;
+    }
+
+    // Check for confirmation flag
+    bool sure = false;
+    cmd_getval(cmdmap, "yes_i_really_mean_it", sure);
+    if (!sure) {
+      ss << "this is a potentially disruptive optionw, clients need to use IDs with access to new fs name.  "
+        "Add --yes-i-really-mean-it if you are sure you wish to continue.";
```
Comment (batrick) [inline src/mon/FSCommands.cc:1008]:
```suggestion
      ss << "this is a potentially disruptive operation, clients cephx credentials need reauthorized to access the file system and its pools with the new name.  "
        "Add --yes-i-really-mean-it if you are sure you wish to continue.";
```

File: src/mds/MDSMap.h [hunk 3]
```diff
@@ -205,6 +205,7 @@ class MDSMap {
   void clear_flag(int f) { flags &= ~f; }
 
   std::string_view get_fs_name() const {return fs_name;}
+  void set_fs_name(std::string new_fs_name) { fs_name = new_fs_name; }
```
Comment (batrick) [inline src/mds/MDSMap.h:208]:
```suggestion
  void set_fs_name(std::string new_fs_name) { fs_name = std::move(new_fs_name); }
```

nit ^

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise code looks okay

**Comment by vshankar** [comment]
@ajarr @batrick Thinking how this change would effect the mirror daemon especially when the mirror daemon is active. I'm guessing the FSMap is updated. In that case the mirror daemon would get a FSMap update and might try to synchronize the file system when it is already synchronizing it based on its old name.

**Comment by batrick** [comment]
@ajarr Don't forget docs (and PendingREleaseNote).

**Comment by batrick** [comment]
> @ajarr @batrick Thinking how this change would effect the mirror daemon especially when the mirror daemon is active. I'm guessing the FSMap is updated. In that case the mirror daemon would get a FSMap update and might try to synchronize the file system when it is already synchronizing it based on its old name.

We could add a requirement to disable mirroring?

File: src/mds/MDSMap.h (hunk 3 above)
Comment (ajarr) [reply src/mds/MDSMap.h]: Done

File: src/mon/FSCommands.cc (hunk 2 above)
Comment (ajarr) [reply src/mon/FSCommands.cc]: Sorry! didn't mean to be tardy. I uploaded the draft just to show my progress.

**Comment by ajarr** [comment]
> > @ajarr @batrick Thinking how this change would effect the mirror daemon especially when the mirror daemon is active. I'm guessing the FSMap is updated. In that case the mirror daemon would get a FSMap update and might try to synchronize the file system when it is already synchronizing it based on its old name.
> 
> We could add a requirement to disable mirroring?

Will do.

**Comment by ajarr** [comment]
> @ajarr Don't forget docs (and PendingREleaseNote).

Thanks! I'll keep this in mind. To be clear, I put up the draft to show my progress. This is not the version for review.

**Comment by vshankar** [comment]
> We could add a requirement to disable mirroring?

We could probably handle this is cephfs-mirror. See: src/tools/cephfs_mirror/ClusterWatcher.cc::handle_fsmap() -- if the fscid is already marked as mirroring enabled but the fs_name mismatches, that indicates a file system has been renamed.

**Comment by vshankar** [comment]
> We could probably handle this is cephfs-mirror. See: src/tools/cephfs_mirror/ClusterWatcher.cc::handle_fsmap() -- if the fscid is already marked as mirroring enabled but the fs_name mismatches, that indicates a file system has been renamed.

BTW, ClusterWatcher should be able to detect the rename and dispatch a "disable mirroring" callback followed by "enable mirroring".

**Comment by ajarr** [comment]
Rename succeeds
```
$ ./bin/ceph fs rename a b --yes-i-really-mean-it
$ ./bin/ceph fs ls
name: b, metadata pool: cephfs.a.meta, data pools: [cephfs.a.data ]
```

Mounting using auth ID with MON and MDS caps specific to old fsname fails as expected. 
```
$ ./bin/ceph auth get client.fs_a
[client.fs_a]
	key = AQBAMYxg8Ao7OhAATEw0kKfiHm6Qu7xn5UnrMA==
	caps mds = "allow rwp fsname=a"
	caps mon = "allow r fsname=a"
	caps osd = "allow rw tag cephfs data=a"
exported keyring for client.fs_a

$ sudo ./bin/ceph-fuse --id=fs_a --client_fs=b /mnt/ceph-fuse/
2021-04-30T12:51:04.853-0400 7f32a4d8c080 -1 init, newargv = 0x5602a25bc140 newargc=15
ceph-fuse[321136]: starting ceph client
ceph-fuse[321136]: ceph mount failed with (2) No such file or directory
2021-04-30T12:51:04.871-0400 7f32a4d8c080 -1 client.4167 mdsmap subscription failed: (2) No such file or directory
```

Mounting using auth ID that has the MON and MDS caps to the new fsname also fails, unexpectedly. 
```
$ ./bin/ceph auth get-or-create client.fs_b mon "allow r fsname=b" mds "allow rwp fsname=b" osd "allow rw tag cephfs data=a"

$ sudo ./bin/ceph-fuse --id=fs_b --client_fs=b /mnt/ceph-fuse/
ceph-fuse[321174]: starting ceph client
2021-04-30T12:51:20.000-0400 7fa7f65e8080 -1 init, newargv = 0x560a03a8a140 newargc=15
2021-04-30T12:51:20.014-0400 7fa7d1ffb640 -1 client.4169 mds.0 rejected us (client doesn't have caps for FS "a")
ceph-fuse[321174]: ceph mount failed with (1) Operation not permitted
```

The failure happens when MDS checks the client session's caps in Server::handle_client_session, where the session's MDS caps is checked with MDSRank's fs_name, which is still the file system's old name.
https://github.com/ceph/ceph/blob/pacific/src/mds/Server.cc#L546
 
 I need to figure out how to also update the MDSRank's fs_name when the `fs rename` command is issued?

**Comment by batrick** [comment]
> I need to figure out how to also update the MDSRank's fs_name when the `fs rename` command is issued?
> https://github.com/ceph/ceph/blob/pacific/src/mds/MDSRank.h#L443

Yup, looks like it.

**Comment by ajarr** [comment]
> > I need to figure out how to also update the MDSRank's fs_name when the `fs rename` command is issued?
> > https://github.com/ceph/ceph/blob/pacific/src/mds/MDSRank.h#L443
> 
> Yup, looks like it.

Done.

File: src/mds/Server.cc [hunk 4]
```diff
@@ -508,7 +508,7 @@ void Server::handle_client_reclaim(const cref_t<MClientReclaim> &m)
     return;
   }
 
-  std::string_view fs_name = mds->get_fs_name();
+  std::string_view fs_name = mds->mdsmap->get_fs_name();
```
Comment (ajarr) [inline src/mds/Server.cc:511]: Patrick, I will add an another PR as you suggested that removes the fs_name from MDSRank;  set map_fs_name in the MMDSMap message to null along with the comment that this was accidentally added when we thought we needed to pass fsid in the messages.

**Comment by ajarr** [comment]
> > We could probably handle this is cephfs-mirror. See: src/tools/cephfs_mirror/ClusterWatcher.cc::handle_fsmap() -- if the fscid is already marked as mirroring enabled but the fs_name mismatches, that indicates a file system has been renamed.
> 
> BTW, ClusterWatcher should be able to detect the rename and dispatch a "disable mirroring" callback followed by "enable mirroring".

Took the simpler approach of just failing to rename if the mirroring is enabled on the file system.

**Comment by ajarr** [comment]
Please review https://github.com/ceph/ceph/pull/41385 that supersedes this PR
I will close this one once that PR looks OK

**Comment by ajarr** [comment]
This

**Pull request closed without merging**